src/Groups/action.jl

```diff
@@ -0,0 +1,237 @@
+#############################################################################
+##
+##  common actions of group elements
+##
+##  The idea is to delegate the action of `GAPGroupElem` objects
+##  on `GAP.GapObj` objects to the corresponding GAP action,
+##  and to implement the action on native Julia objects case by case.
+
+export on_points, on_right, on_tuples, on_sets, permuted
+
+
+"""
+    on_points(pnt::GAP.GapObj, x::GAPGroupElem)
+    on_points(pnt::GAPGroupElem, x::GAPGroupElem)
+    on_points(pnt::Int, x::PermGroupElem)
+
+Return the image of `pnt` under `x`.
+If `pnt` is a `GAP.GapObj` then the action is given by `GAP.Globals.OnPoints`.
+If `pnt` is a `GAPGroupElem` then the action is given by conjugation,
+that is, the result is `x`^-1 * `pnt` * `x`.
+If `pnt` is a positive integer and `x` is a permutation then the action
+is given by the natural action that maps `pnt` to `x(pnt)`.
+
+# Examples
+```jldoctest
+julia> g = symmetric_group(3);  g[1]
+(1,2,3)
+
+julia> g[2]
+(1,2)
+
+julia> on_points(g[2], g[1])
+(2,3)
+
+julia> on_points(g[2].X, g[1])
+GAP: (2,3)
+
+julia> on_points(1, g[1])
+2
+
+```
+"""
+on_points(pnt::GAP.GapObj, x::GAPGroupElem) = GAP.Globals.OnPoints(pnt, x.X)
+
+on_points(pnt::GAPGroupElem, x::GAPGroupElem) = inv(x) * pnt * x
+
+on_points(pnt::Int, x::PermGroupElem) = GAP.Globals.OnPoints(pnt, x.X)
+
+
+"""
+    on_right(pnt::GAP.GapObj, x::GAPGroupElem)
+    on_right(pnt::GAPGroupElem, x::GAPGroupElem)
+    on_right(v::Vector{T}, x::MatrixGroupElem) where T
+
+Return the image of the point `pnt` under `x`,
+where the action is given by right multiplication with `x`.
+
+# Examples
+```jldoctest
+julia> g = general_linear_group(2,3);  mat = g[1]
+[ [ Z(3), 0*Z(3) ], [ 0*Z(3), Z(3)^0 ] ]
+
+julia> on_right(mat, mat)
+[ [ Z(3)^0, 0*Z(3) ], [ 0*Z(3), Z(3)^0 ] ]
+
+julia> r = mat.X[1]
+GAP: [ Z(3), 0*Z(3) ]
+
+julia> on_right( r, mat )
+GAP: [ Z(3)^0, 0*Z(3) ]
+
+julia> v = Vector{GAP.FFE}(r)
+2-element Array{FFE,1}:
+ GAP: Z(3)
+ GAP: 0*Z(3)
+
+julia> on_right(v, mat)
+2-element Array{GAP.FFE,1}:
+ GAP: Z(3)^0
+ GAP: 0*Z(3)
+
+```
+"""
+on_right(pnt::GAP.GapObj, x::GAPGroupElem) = GAP.Globals.OnRight(pnt, x.X)
+
+on_right(pnt::GAPGroupElem, x::GAPGroupElem) = pnt * x
+
+function on_right(v::Vector{T}, x::MatrixGroupElem) where T
+    gapv = GAP.julia_to_gap(v)
+    img = GAP.Globals.OnRight(gapv, x.X)
+    return Vector{T}(img)
+end
+
+
+"""
+    on_tuples(tuple::GAP.GapObj, x::GAPGroupElem)
+    on_tuples(tuple::Vector{T}, x::GAPGroupElem) where T
+    on_tuples(tuple::T, x::GAPGroupElem) where T <: Tuple
```

Comment:
How about this instead:
```suggestion
    on_tuples(tuple::Tuple{Vararg{T}}, x::GAPGroupElem) where T
```

Going beyond this (and perhaps beyond this PR): as mentioned already, for tuples, it'd also be natural to just define a `^` method directly, as it is "clear" how to act on that. Just as I'd argue that the default action for a vector ought to be OnRight, for a `Set` or `BitSet` OnSets etc. -- and then nested, i.e. a `Set{Set{Int}}` suggest `OnSetsSets`, while `Set{Tuple{Vararg{Int8}}}` suggests `OnSetsTuples` etc.

Reply:
In my variant, the tuple need not be homogeneous, and it is simpler to create the result object of the right type.